src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,514 @@
+
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Generic
+{
+    /// <summary>
+    /// Represents a data structure in which each element additionally has a priority
+    /// associated with it.
+    /// </summary>
+    /// <typeparam name="TElement">The type of the element.</typeparam>
+    /// <typeparam name="TPriority">The type of the priority.</typeparam>
+    public class PriorityQueue<TElement, TPriority>
+    {
+        /// <summary>
+        /// Represents an implicit heap-ordered complete d-ary tree, stored as an array.
+        /// </summary>
+        private readonly List<(TElement element, TPriority priority)> nodes;
+
+        private const int RootIndex = 0;
+
+        /// <summary>
+        /// Specifies the arity of the d-ary heap, which here is quaternary.
+        /// </summary>
+        private const int Arity = 4;
+
+        /// <summary>
+        /// Creates an empty priority queue.
+        /// </summary>
+        public PriorityQueue()
+        {
+            this.nodes = new List<(TElement, TPriority)>();
+            this.UnorderedItems = new UnorderedItemsCollection(this.nodes);
+            this.Comparer = Comparer<TPriority>.Default;
+        }
+
+        /// <summary>
+        /// Creates an empty priority queue with the specified initial capacity for its underlying array.
+        /// </summary>
+        public PriorityQueue(int initialCapacity)
+        {
+            if (initialCapacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(initialCapacity));
+            }
+
+            this.nodes = new List<(TElement, TPriority)>(initialCapacity);
+            this.UnorderedItems = new UnorderedItemsCollection(this.nodes);
+            this.Comparer = Comparer<TPriority>.Default;
+        }
+
+        /// <summary>
+        /// Creates an empty priority queue with the specified priority comparer.
+        /// </summary>
+        public PriorityQueue(IComparer<TPriority>? comparer)
+        {
+            this.nodes = new List<(TElement, TPriority)>();
+            this.UnorderedItems = new UnorderedItemsCollection(this.nodes);
+            this.Comparer = comparer ?? Comparer<TPriority>.Default;
+        }
+
+        /// <summary>
+        /// Creates an empty priority queue with the specified priority comparer and
+        /// the specified initial capacity for its underlying array.
+        /// </summary>
+        public PriorityQueue(int initialCapacity, IComparer<TPriority>? comparer)
+        {
+            if (initialCapacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(initialCapacity));
+            }
+
+            this.nodes = new List<(TElement, TPriority)>(initialCapacity);
+            this.UnorderedItems = new UnorderedItemsCollection(this.nodes);
+            this.Comparer = comparer ?? Comparer<TPriority>.Default;
+        }
```

Comment:
There's a lot of duplication with the above ctors.  It'd be better to pick one or two and have the rest delegate to them.  That does mean we might end up with an extra null check or the like here or there in the ctors (e.g. checking whether comparer is null even though we know it's not)... that's ok.